File: go/usage.go
// this file is @generated
package meteroid

import (
	"context"
	"net/http"
)

// UsageGetCustomerUsageOptions carries the query and header parameters of
// Usage.GetCustomerUsage.
//
// Optional parameters are pointers; leave them nil to omit them.
type UsageGetCustomerUsageOptions struct {
	StartDate string

	EndDate string

	MetricId *BillableMetricId
}

// UsageGetSubscriptionUsageOptions carries the query and header parameters of
// Usage.GetSubscriptionUsage.
//
// Optional parameters are pointers; leave them nil to omit them.
type UsageGetSubscriptionUsageOptions struct {
	StartDate *string

	EndDate *string

	MetricId *BillableMetricId
}

// UsageGetUsageSummaryOptions carries the query and header parameters of
// Usage.GetUsageSummary.
//
// Optional parameters are pointers; leave them nil to omit them.
type UsageGetUsageSummaryOptions struct {
	StartDate string

	EndDate string

	MetricId *BillableMetricId
}

// Usage groups the usage operations of the Meteroid API.
type Usage struct {
	client *Client
}

// Retrieve aggregated usage data for a customer over a specified period.
func (a *Usage) GetCustomerUsage(ctx context.Context, customerId string, options UsageGetCustomerUsageOptions) (*UsageResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/usage/customer/{customer_id}")
	req.SetPathParam("customer_id", customerId)

	req.SetQueryParam("start_date", options.StartDate)
	req.SetQueryParam("end_date", options.EndDate)
	if options.MetricId != nil {
		req.SetQueryParam("metric_id", string(*options.MetricId))
	}

	var out UsageResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Retrieve aggregated usage data for a subscription's usage-based components.
// If start_date/end_date are omitted, defaults to the current billing period.
func (a *Usage) GetSubscriptionUsage(ctx context.Context, subscriptionId string, options *UsageGetSubscriptionUsageOptions) (*UsageResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/usage/subscription/{subscription_id}")
	req.SetPathParam("subscription_id", subscriptionId)

	if options != nil {
		if options.StartDate != nil {
			req.SetQueryParam("start_date", *options.StartDate)
		}
		if options.EndDate != nil {
			req.SetQueryParam("end_date", *options.EndDate)
		}
		if options.MetricId != nil {
			req.SetQueryParam("metric_id", string(*options.MetricId))
		}
	}

	var out UsageResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Retrieve aggregated usage data across all customers for the tenant.
func (a *Usage) GetUsageSummary(ctx context.Context, options UsageGetUsageSummaryOptions) (*UsageResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/usage/summary")

	req.SetQueryParam("start_date", options.StartDate)
	req.SetQueryParam("end_date", options.EndDate)
	if options.MetricId != nil {
		req.SetQueryParam("metric_id", string(*options.MetricId))
	}

	var out UsageResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: python/meteroid/api/usage.py
# this file is @generated
import typing as t

from ..models import (
    BillableMetricId,
    UsageResponse,
)
from .common import ApiBaseAsync, ApiBaseSync, serialize_query_params


class UsageAsync(ApiBaseAsync):
    """usage API."""

    async def get_customer_usage(
        self,
        customer_id: str,
        *,
        start_date: str,
        end_date: str,
        metric_id: t.Optional[BillableMetricId] = None,
    ) -> UsageResponse:
        """Retrieve aggregated usage data for a customer over a specified period."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/usage/customer/{customer_id}",
            path_params={
                "customer_id": customer_id,
            },
            query_params=serialize_query_params(
                {
                    "start_date": start_date,
                    "end_date": end_date,
                    "metric_id": metric_id,
                },
            ),
        )
        return UsageResponse.from_dict(response.json())

    async def get_subscription_usage(
        self,
        subscription_id: str,
        *,
        start_date: t.Optional[str] = None,
        end_date: t.Optional[str] = None,
        metric_id: t.Optional[BillableMetricId] = None,
    ) -> UsageResponse:
        """Retrieve aggregated usage data for a subscription's usage-based components.
        If start_date/end_date are omitted, defaults to the current billing period."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/usage/subscription/{subscription_id}",
            path_params={
                "subscription_id": subscription_id,
            },
            query_params=serialize_query_params(
                {
                    "start_date": start_date,
                    "end_date": end_date,
                    "metric_id": metric_id,
                },
            ),
        )
        return UsageResponse.from_dict(response.json())

    async def get_usage_summary(
        self,
        *,
        start_date: str,
        end_date: str,
        metric_id: t.Optional[BillableMetricId] = None,
    ) -> UsageResponse:
        """Retrieve aggregated usage data across all customers for the tenant."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/usage/summary",
            query_params=serialize_query_params(
                {
                    "start_date": start_date,
                    "end_date": end_date,
                    "metric_id": metric_id,
                },
            ),
        )
        return UsageResponse.from_dict(response.json())


class Usage(ApiBaseSync):
    """usage API."""

    def get_customer_usage(
        self,
        customer_id: str,
        *,
        start_date: str,
        end_date: str,
        metric_id: t.Optional[BillableMetricId] = None,
    ) -> UsageResponse:
        """Retrieve aggregated usage data for a customer over a specified period."""
        response = self._request_sync(
            method="get",
            path="/api/v1/usage/customer/{customer_id}",
            path_params={
                "customer_id": customer_id,
            },
            query_params=serialize_query_params(
                {
                    "start_date": start_date,
                    "end_date": end_date,
                    "metric_id": metric_id,
                },
            ),
        )
        return UsageResponse.from_dict(response.json())

    def get_subscription_usage(
        self,
        subscription_id: str,
        *,
        start_date: t.Optional[str] = None,
        end_date: t.Optional[str] = None,
        metric_id: t.Optional[BillableMetricId] = None,
    ) -> UsageResponse:
        """Retrieve aggregated usage data for a subscription's usage-based components.
        If start_date/end_date are omitted, defaults to the current billing period."""
        response = self._request_sync(
            method="get",
            path="/api/v1/usage/subscription/{subscription_id}",
            path_params={
                "subscription_id": subscription_id,
            },
            query_params=serialize_query_params(
                {
                    "start_date": start_date,
                    "end_date": end_date,
                    "metric_id": metric_id,
                },
            ),
        )
        return UsageResponse.from_dict(response.json())

    def get_usage_summary(
        self,
        *,
        start_date: str,
        end_date: str,
        metric_id: t.Optional[BillableMetricId] = None,
    ) -> UsageResponse:
        """Retrieve aggregated usage data across all customers for the tenant."""
        response = self._request_sync(
            method="get",
            path="/api/v1/usage/summary",
            query_params=serialize_query_params(
                {
                    "start_date": start_date,
                    "end_date": end_date,
                    "metric_id": metric_id,
                },
            ),
        )
        return UsageResponse.from_dict(response.json())
